### ISSUE TITLE
[Usage]: Running LLMEngine (or AsyncLLMEngine) in Arrow Flight RPC do_exchange()

### ISSUE THREAD

**Issue by pvardanis** [vllm-project/vllm#8487]
### Your current environment

```text
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: macOS 14.2.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.*******)
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.19 (default, May  8 2024, 17:10:30)  [Clang 15.0.0 (clang-1500.*******)] (64-bit runtime)
Python platform: macOS-14.2.1-arm64-arm-64bit
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] mypy==1.7.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.4
[pip3] pyzmq==25.1.2
[conda] Could not collect
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: N/A
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
Could not collect
```

### How would you like to use vllm

I want to run inference for any LLM model using `LLMEngine` (or `AsyncLLMEngine`) and I'm using [Arrow Flight RPC](https://arrow.apache.org/docs/python/flight.html) as a server. In particular, I'm using the [do_exchange](https://arrow.apache.org/docs/python/generated/pyarrow.flight.FlightServerBase.html#pyarrow.flight.FlightServerBase.do_exchange) method to serve model inference.

This RPC is a bi-directional channel where both client and server have access to a reader and a writer and can read/write from/to the channel at the same time. The problem is that Arrow Flight doesn't expose `gRPC`'s async API and uses multi-threading instead of async. So far, for serving other types of models from multiple clients requests arriving at the same time it does so by spawning up multiple threads and running in parallel. I'm using a [threading.Lock()](https://docs.python.org/3/library/threading.html#threading.Lock) mechanism where I'm acquiring a lock before calling model inference and releasing it afterwards to make sure everything is processed securely in case of thread unsafe code. 

It has worked great so far for other types of ML frameworks (e.g. tensorflow, PyTorch, Hugging Face etc.) and now looking for ways to integrate vLLM to take advantage of continuous batching. I've already tested the `LLM` class (that uses `LLMEngine` under the hood) and it has worked great, however I'm not getting advantage of continuous batching at all. I'm locking `LLM.generate()` to which I pass a batch of input prompts and have to wait for all requests in the batch to finish before releasing the lock and finally writing to the bi-directional channel, instead of writing the results in the channel one-by-one as soon as they're finished. This is ideal for offline/batch inference and also for scenarios under dynamic batching (which I've implemented successfully on the server side), but not good for real-time scenarios where single prompts arrive from multiple users.

Due to the sync nature of the API I cannot use `AsyncLLMEngine` or at least I don't know how to or if it's gonna work, mainly because it's not a best practice to run an event loop in a multi-threaded environment.

Here's a rough implementation of the model:
```python
class VLLMInference:
    def __init__(self) -> None:
        self.model = LLM(
            model=f"{model_path}/artifacts/Meta-Llama-3-8B-Instruct/"
        )

    def predict(self, input_data):
        prompts = input_data["text"].tolist()

        sampling_params = SamplingParams(temperature=0.1, top_p=0.95)
        responses = self.model.generate(prompts, sampling_params)

        return {"generated_text": np.array([response.outputs[0].text for response in responses])}
```

and the server:

```python
def do_exchange(
        self,
        context: flight.ServerCallContext,
        descriptor: flight.FlightDescriptor,
        reader: flight.FlightStreamReader,
        writer: flight.FlightStreamWriter,
    ) -> None:
        """This method implements the `do_exchange` method of the FlightServerBase
        class.

        :param context: A ServerCallContext object.
        :param descriptor: A FlightDescriptor object.
        :param reader: A FlightStreamReader object.
        :param writer: A FlightStreamWriter object.
        """
        is_first_batch = True
        while True:
            logger.info("Processing data...") # leaving the channel open
            (
                writer,
                reader,
                is_first_batch,
            ) = self._process_data(writer, reader, is_first_batch)
            logger.info("Output data ready to be consumed.")

def _process_data(
        self,
        batch: pa.RecordBatch,
        writer: flight.MetadataRecordBatchWriter,
        is_first_batch: bool,
    ) -> Tuple[flight.FlightStreamWriter, bool]:
        # some code that acquires the lock
        result = self.inference.predict(batch)
        # some code that releases the lock
        return self._write_result(writer, result, is_first_batch)  # type: ignore

def _write_result(
        self,
        writer: flight.FlightStreamWriter,
        result: List[pa.RecordBatch],
        is_first_batch: bool,
    ) -> Tuple[flight.FlightStreamWriter, bool]:
        if is_first_batch is True:
            is_first_batch = False
            logger.debug("Writing schema to stream...")
            writer.begin(result[0].schema)

        for batch in result:
            writer.write_batch(batch)

        return (writer, is_first_batch)
```

UPDATE: I've also implemented a "hacky" async version of my server, that unfortunately is mixing the sync functionality of `do_exchange` with supporting async functions.

```python
def do_exchange( # we have no control over this
        self,
        context: flight.ServerCallContext,
        descriptor: flight.FlightDescriptor,
        reader: flight.FlightStreamReader,
        writer: flight.FlightStreamWriter,
    ) -> None:
        """This method implements the `do_exchange` method of the FlightServerBase
        class.

        :param context: A ServerCallContext object.
        :param descriptor: A FlightDescriptor object.
        :param reader: A FlightStreamReader object.
        :param writer: A FlightStreamWriter object.
        """
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)
        is_first_batch = True
        while True:
            logger.info("Processing data...") # leaving the channel open
            (
                writer,
                reader,
                is_first_batch,
            ) = loop.run_until_complete(self._process_data(writer, reader, is_first_batch))
            logger.info("Output data ready to be consumed.")

async def _process_data(
        self,
        batch: pa.RecordBatch,
        writer: flight.MetadataRecordBatchWriter,
        is_first_batch: bool,
    ) -> Tuple[flight.FlightStreamWriter, bool]:
        # lock is not used at this point, but inside predict instead
        result = await self.inference.predict(batch)
        return self._write_result(writer, result, is_first_batch)
```

that allows me to have async `Inference` classes. For `VLLMInference` I'm thinking of something like this:

```python
class VLLMInference:
    def __init__(self) -> None:
        engine_args = AsyncEngineArgs(
            model=model=f"{model_path}/artifacts/Meta-Llama-3-8B-Instruct/"
        )
        self.model = AsyncLLMEngine.from_engine_args(engine_args)
        self._lock = asyncio.Lock()

    async def predict(self, input_data):
        prompt = input_data["text"].tolist()[0]
        sampling_params = SamplingParams(temperature=0.1, top_p=0.95)
        request_id = random_uuid()
        
        results_generator = engine.generate(prompt, sampling_params, request_id)

        async with self._lock: # make sure it's not accessed by multiple threads since gRPC is multi-threaded
            async for request_output in results_generator:
                 final_output = request_output

        return {"generated_text": final_output.outputs[0].text)}
```
Does it make sense? The main thing that worries me is that the app is multi-threaded and I've seen that at least the `LLM.generate()` is not thread-safe and only using the lock fixed the [issues](https://github.com/vllm-project/vllm/issues/4556) I was getting. That's why I'm also putting a lock when reading the outputs from the generator, to make sure no more than one of the server threads (i.e. each thread serves a client if both arrive at almost the same time) access `engine.step()` at the same time. I'm just wondering whether that would limit the support for "actual" continuous batching or if it makes sense at all in the context of the hybrid multi-threaded/async environment I'm forced to use.
